### PR TITLE
Don't create a new languageservice on every build + incremental builds

### PR DESCRIPTION
**Pull request by Matsuuu**
Heya,

As per our discussion here and on slack, I've fiddled around with the code base a bit.

I did a simple map-based "cache" for the typescript builders, saving them by project id's and rewrote the way we handled the languageservices.

I made the now cached instance of TypescriptBuilder save the languageService and the languageServiceHost as class level properties, so that we can re-use them instead of instantiating a new one every build.

I also expanded the WorkerLanguageServiceHost class to handle the new way of adding, and syncing the project files. Now with the added versioning, we can just update the necessary files, and by bumping up the version number on those files, the languageservice will be able to incrementally build them.

A lot of inspiration was drawn from the [TS docs](https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API#incremental-build-support-using-the-language-services). I'd love to have a review on this and if this is going in the right direction?

From my local testing, everything should work as it did before, just a bit snappier and more efficiently.

If this change and coding style is fine, I'd love to delve deeper into the actual language service support and integrating more LSP features to the playground.

### CODE REVIEW

File: src/typescript-worker/project-cache.ts
Comment (aomarks) [inline src/typescript-worker/project-cache.ts:8]: Related to other comment -- since we know that there's a 1-1 relationship between each `<playground-project>` and its dedicated web worker, I think we don't actually need this cache. We can just make one module-scoped TypeScript builder :D

File: src/typescript-worker/typescript.ts
Comment (aomarks) [inline src/typescript-worker/typescript.ts:52]: Maybe simpler to do this initialization on construction using a class field initializer? E.g.:

```ts
  private readonly _languageServiceHost = new WorkerLanguageServiceHost(
    self.origin,
    compilerOptions
  );

  private readonly _languageService = ts.createLanguageService(
    this._languageServiceHost,
    ts.createDocumentRegistry()
  );
```

File: src/typescript-worker/typescript.ts
Comment (aomarks) [inline src/typescript-worker/typescript.ts:220]: What happens if a file is deleted? Maybe we should iterate over the current `files`, and remove the ones that don't exist anymore?

I wonder if that will work properly in this edge case?:

1. compile with foo at version 0
2. delete foo and compile without it
3. re-create foo with different content, but it gets version 0 again because its new

**Review by aomarks — CHANGES_REQUESTED** [review]
This is looking fantastic! This should improve compilation performance a lot right away, and sets things up nicely to be able to query it directly for completions etc.!

I have a few higher level comments, and then I'll do another round with some very minor style stuff.

But generally looking great! I think we should merge this as an isolated change once it's ready, since it already should have a big benefit.

File: src/playground-project.ts
Comment (Matsuuu) [reply src/playground-project.ts:548]: Ah yeah that makes sense. I didn't have the full image of the project, and which layers are shared and which have their own instance.

If this is the case, I'll just remove the sessionid param and modify the caching into just a module scoped instance

File: src/typescript-worker/typescript.ts
Comment (Matsuuu) [reply src/typescript-worker/typescript.ts:52]: You're totally right. I did pass the files into the function earlier, but as I removed that and moved the file passing elsewhere, this can now be done in the constructor.

File: src/typescript-worker/typescript.ts
Comment (Matsuuu) [reply src/typescript-worker/typescript.ts:220]: Oh good point! I'll need to cover the deletion case. 

.. What comes to the deletion and re-addition, I'll need to check out if the LS is capable of handling it out of the box, or if we need to add some flags etc.

**Comment by Matsuuu** [comment]
Cheers on the review! I implemented the changes proposed, and made the languageservice aware of removed files. I tested it out with re-naming, re-creating the file, etc. and didn't notice any hiccups.

If we merge this, I'll next start working on building the LSP functionalities

File: src/typescript-worker/typescript.ts
Comment (aomarks) [inline src/typescript-worker/typescript.ts:209]:
```suggestion
  private _removeDeletedFiles(files: Map<string, string>) {
```

File: src/typescript-worker/typescript.ts
Comment (aomarks) [inline src/typescript-worker/typescript.ts:217]:
```suggestion
  private _addFile(fileName: string, file: VersionedFile) {
```

Could also just consider moving its body directly into `updateFileContentIfNeeded`, because it's not used anywhere else (at least, after my other suggestion to call `updateFileContentIfNeeded` instead of `addFile` for the `d.ts` files).

File: src/typescript-worker/typescript.ts
Comment (aomarks) [inline src/typescript-worker/typescript.ts:227]:
```suggestion
  private _removeFile(fileName: string) {
```

Again, could also consider just moving the body directly into `updateFileContentIfNeeded` since it's only used in one place, and is very simple self-describing code already.

File: src/typescript-worker/typescript.ts
Comment (aomarks) [inline src/typescript-worker/typescript.ts:231]:
```suggestion
  private _fileContentHasUpdated(oldFile: VersionedFile, newContent: string) {
```

Here too -- it's pretty simple self-documenting code only used in one place, so you could just do the check directly in `updateFileContentIfNeeded`.

File: src/typescript-worker/build.ts
Comment (aomarks) [inline src/typescript-worker/build.ts:16]: Very minor, but mind running `npm run format` to Prettier-format everything?

**Review by aomarks — CHANGES_REQUESTED** [review]
Nice! Looking good. Two comments that affect the behavior, everything else just some style suggestions.

Would you also mind adding an entry to the `CHANGELOG.md` file? Something like

"TypeScript files are now compiled incrementally for improved performance."

**Comment by aomarks** [comment]
> Would you also mind adding an entry to the `CHANGELOG.md` file? Something like

Oh also I just recently merged https://github.com/PolymerLabs/playground-elements/pull/204 which upgrades to Lit 2 and updates the CHANGELOG, so you might want to `merge main` before you modify the CHANGELOG, because otherwise we'll probably have to do a merge conflict resolve afterwards.

File: src/typescript-worker/project-cache.ts
Comment (Matsuuu) [reply src/typescript-worker/project-cache.ts:6]: If we are going to implement more LSP support (completion, types, etc.), we need a LS instance for those too.

I was wondering if we should/could use the same `TypeScriptBuilder` instance for those too, since it will be up to date with the latest project files? If we do that, we might want to keep the `project-cache.ts` separate, and maybe even re-name the builder to something else.

Thoughts?

Comment (aomarks) [reply src/typescript-worker/project-cache.ts:6]: Good point, it will probably make sense to have some separation for the reason you said. And yeah, I also agree it will probably make sense to rename it -- or maybe we'll want to factor out another class that purely encapsulate the language service parts, which is passed *in* to the builder. Something to consider in the near future!

Only thing we should do is add a license header to this file, then:

```ts
/**
 * @license
 * Copyright 2021 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
```

**Comment by Matsuuu** [comment]
Cheers on the thorough review once again @aomarks !

I've done the changes, applied formatting and coding styles accordingly, and removed reduntant functions.

I also merged the upstream main and updated the changelog. 

After we merge this, I'll start working on a plan on re-structuring the typescriptworker for the LSP functionality (in a different branch / pr of course).

**Review by aomarks — APPROVED** [review]

**Comment by aomarks** [comment]
A few simple lint errors in https://github.com/PolymerLabs/playground-elements/pull/205/checks?check_run_id=3653621020 to fix, then ready to merge.

**Comment by Matsuuu** [comment]
Ahh duh. Of course I forgot to lint :D 

Fixed it now!

**Comment by aomarks** [comment]
> Ahh duh. Of course I forgot to lint :D
> 
> Fixed it now!

No worries.

I think the formatting has reverted again, indentation looks different, could you also just run `npm run format`?

**Comment by Matsuuu** [comment]
Sorry for the inconvenience. As I said earlier, still kinda new to these larger projects and contributing :) Should be fine now

**Comment by aomarks** [comment]
> Sorry for the inconvenience. As I said earlier, still kinda new to these larger projects and contributing :) Should be fine now

No problem!

**Pull request merged**
